init.lua

```diff
@@ -113,66 +113,58 @@ function Highlight:tokenize_line(idx, state)
 	if not self.doc.treesit then return oldTokenize(self, idx, state) end
 
 	local res = {}
-	local i = idx - 1
 	res.init_state = state
 	res.text = self.doc.lines[idx]
 	res.state = 0
 	res.tokens = {}
 
-	local linenodes = {}
-	local gotline = false
-	print(idx)
-	local lastNode
+	local i = idx - 1
+	local tokens = res.tokens
+	local currentLine = self.doc.lines[idx]
+	local lastNode, lastStartPoint, lastEndPoint
 	for n, nName in self.doc.ts.query:capture(self.doc.ts.tree:root()) do
-		lastNode = (#linenodes ~= 0 and linenodes[#linenodes] or {})['node']
 		local startPoint = n:start_point()
 		local endPoint = n:end_point()
 
-		if startPoint.row == i and endPoint.row == i then
-			-- indentation pass
-			if (startPoint.column > 1 or self.doc.lines[idx]:sub(0, startPoint.column):find('\t')) and gotline == false then
-				table.insert(res.tokens, 'normal')
-				table.insert(res.tokens, self.doc.lines[idx]:sub(0, startPoint.column))
+		if i > startPoint.row and i > endPoint.row then goto continue end
+		if startPoint.row > i then break end
+
+		-- single line token
+		if startPoint.row == endPoint.row then
+			if not lastNode and startPoint.column > 0 then
+				-- first node
+				tokens[#tokens+1] = 'normal'
+				tokens[#tokens+1] = currentLine:sub(1, startPoint.column)
+			elseif lastNode and lastEndPoint.row == startPoint.row and startPoint.column - lastEndPoint.column > 0 then
+				-- node from the same line and have spaces between them
+				tokens[#tokens+1] = 'normal'
+				tokens[#tokens+1] = currentLine:sub(lastEndPoint.column + 1, startPoint.column)
 			end
 
-			if lastNode then
-				local lnStartPoint = lastNode:start_point()
-				local lnEndPoint = lastNode:end_point()
-
-				if lnStartPoint.column == startPoint.column and lnEndPoint.column == endPoint.column then
-					linenodes[#linenodes] = {node = n, name = nName}
-					res.tokens[#res.tokens - 1] = nName
-					res.tokens[#res.tokens] = n:source()
-					goto continue
-				end
-				-- whitespace between cur and last node
-				if startPoint.column - lnEndPoint.column >= 1 then
-					table.insert(res.tokens, 'normal')
-					-- todo: use whitespace from line
-					table.insert(res.tokens, (' '):rep(startPoint.column - lnEndPoint.column))
-				end
+			local append_idx =
+				(lastNode and lastStartPoint.column == startPoint.column and lastEndPoint.column == endPoint.column) -- if the nodes overlap
+					and (#tokens - 1) -- replace the old one
+					or (#tokens + 1)  -- create a new token
+			tokens[append_idx] = nName
+			tokens[append_idx+1] = n:source()
+
+		elseif i >= startPoint.row and i <= endPoint.row then
+			tokens[#tokens+1] = nName
+			if i == startPoint.row then
+				tokens[#tokens+1] = currentLine:sub(startPoint.column + 1, -2) -- from node start to EOL
```

Comment:
```suggestion
				tokens[#tokens+1] = currentLine:sub((startPoint.column + 1) - currentLine:len() - 2) -- to node start from EOL
```

Reply:
Not too sure why you would want this, its not immediately clear what the thing does (because negative 1st index is rarely used). Perhaps `:sub(startPoint.column + 1)` is what you wanted?

Other than that `(startPoint.column + 1) - currentLine:len() - 2` is off-by-one, it should be subtracted by 1 instead of 2